docs/proposals/20210203-unmanaged-infrastructure.md

```diff
@@ -0,0 +1,236 @@
+
+### Goals
+
+- Introduce support for "externally managed" cluster infrastructure consistently across Cluster API providers.
+- Any machine controller or machine infrastructure controllers must be able to keep operating like they do today.
+- Reuse existing InfraCluster CRDs in "externally managed" clusters to minimise differences between the two topologies.
+
+### Non-Goals/Future Work
+
+- Modify existing managed behaviour.
+- Automatically mark InfraCluster resources as ready (this will be up to the external management component initially).
+- Support anything other than cluster infrastructure (e.g. machines).
+
+## Proposal
+
+A new annotation `cluster.x-k8s.io/managed-by: "<name-of-system>"` is going to be defined in Cluster API core repository, which helps define and identify resources managed by external controllers. The value of the annotation will not be checked by Cluster API and is considered free form text.
+
+Infrastructure providers SHOULD respect the annotation and its contract.
+
+When this annotation is present on an InfraCluster resource, the InfraCluster controller is expected to ignore the resource and not perform any reconciliation.
+Importantly, it will not modify the resource or its status in any way.
+A predicate will be provided in the Cluster API repository to aid provider implmentations in filtering resources that are externally managed.
+
+Additionally, the external management system must provide all required fields within the spec of the InfraCluster and must adhere to the CAPI provider contract and set the InfraCluster status to be ready when it is appropriate to do so.
+
+While an "externally managed" InfraCluster won't reconcile or manage the lifecycle of the cluster infrastructure, CAPI will still be able to create compute nodes within it.
+
+The machine controller must be able to operate without hard dependencies regardless of the cluster infrastructure being managed or externally managed.
+![](https://i.imgur.com/nA61XJt.png)
+
+### User Stories
+
+#### Story 1 - Alternate control plane provisioning with user managed infrastructure
+As a cluster provider I want to use CAPI in my service offering to orchestrate Kubernetes bootstraping while letting workload cluster operators own their infrastructure lifecycle.
+
+#### Story 2 - Restricted access to cloud provider APIs
+As a cluster operator I want to use CAPI to orchestrate kubernetes bootstraping while restricting the privileges I need to grant for my cloud provider because of organisational cloud security constraints.
+
+#### Story 3 - Consuming existing cloud infrastructure
+As a cluster operator I want to use CAPI to orchestate Kubernetes bootstraping while reusing infrastructure that has already been created in the organisation either by me or another team.
+
+### Implementation Details/Notes/Constraints
```

Comment:
Can you please clarify how this annotation will be used for the auto-healing of nodes based on machine health checks?

Reply:
Also, how will the other Infra controllers honor this annotation on the InfraCluster CR?

Reply:
>Can you please clarify how this annotation will be used for the auto-healing of nodes based on machine health checks?

This is transparent for MHC. ```- Any machine controller or machine infrastructure controllers must be able to keep operating like they do today.``` is a goal of this proposal.

>Also, how will the other Infra controllers honor this annotation on the InfraCluster CR?

Which ones in particular are you referring to? only the infraCluster controller is meant to read this annotation.

Reply:
Sorry about the delay in getting back on this. But I was referring to the InfraMachine controllers and also the load balancer controller that will be implemented as part of the loadbalancer provider proposal. But your point of the annotation being consumed only by the InfraCluster controller clarifies. Thanks!